smithy-kotlin-codegen/src/test/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitorTest.kt

```diff
@@ -5,91 +5,261 @@
     }
 
     @Test
-    fun testInputFieldExpression() {
-        testVisit(
-            true,
-            "input.id",
-            "id",
-            "val input = it.input",
-            "val id = input?.id",
+    fun testFunctionExpression_Contains() {
+        assertVisit(
+            "contains(foo, bar)",
+            "contains",
+            "val foo = it?.foo",
+            "val bar = it?.bar",
+            "val contains = foo?.contains(bar) ?: false",
         )
     }
 
     @Test
-    fun testOutputFieldExpression() {
-        testVisit(
-            true,
-            "output.name",
-            "name",
-            "val output = it.output",
-            "val name = output?.name",
+    fun testFunctionExpression_Length() {
+        assertVisit(
+            "length(foo)",
+            "length",
+            "import aws.smithy.kotlin.runtime.util.length",
+            "",
+            "val foo = it?.foo",
+            "val length = foo?.length ?: 0",
         )
     }
 
-    private fun testVisit(includeInput: Boolean, path: String, expectedActualName: String, vararg vals: String) {
-        val visitor = KotlinJmespathExpressionVisitor(
-            includeInput,
-            model,
-            provider,
-            inputShape,
-            inputSymbol,
-            outputShape,
-            outputSymbol
+    @Test
+    fun testIndexExpression() {
+        assertUnimplemented("foo[0]")
+    }
+
+    @Test
+    fun testLiteralExpressions() {
+        assertVisit("`\"a string\"`", "string", "val string = \"a string\"")
+        assertVisit("`3.14`", "number", "val number = 3.14")
+        assertVisit("`false`", "bool", "val bool = false")
+        assertVisit("`null`", "null")
+        assertUnimplemented("`[]`")
+        assertUnimplemented("`{}`")
+    }
+
+    @Test
+    fun testMultiSelectHashExpression() {
+        assertUnimplemented("foo.{bar: baz}")
+    }
+
+    @Test
+    fun testMultiSelectListExpression() {
+        assertVisit(
+            "foo[][bar, baz]",
+            "projection",
+            "val foo = it?.foo",
+            "val fooOrEmpty = foo ?: listOf()",
+            "val projection = fooOrEmpty.flatMap {",
+            "    val multiSelect = listOfNotNull(",
+            "        run {",
+            "            val bar = it?.bar",
+            "            bar",
+            "        },",
+            "        run {",
+            "            val baz = it?.baz",
+            "            baz",
+            "        },",
+            "    )",
+            "    multiSelect",
+            "}",
         )
+    }
 
-        val expression = JmespathExpression.parse(path)
-        expression.accept(visitor)
+    @Test
+    fun testNotExpression() {
+        assertUnimplemented("!foo")
+    }
+
+    @Test
+    fun testObjectProjection() {
+        assertVisit(
+            "foo.*.bar",
+            "projection",
+            "val foo = it?.foo",
+            "val fooValues = foo?.values ?: listOf()",
```

Comment:
where does `.values` come from?

Reply:
That comes from `visitObjectProjection`. It refers specifically to the values of a map, which will then be run through whatever mapping step applies to the projection.

Reply:
Ok so this MUST be a map type that the object expression applies to then?

Reply:
Correct, although the codegen has no way of knowing whether that's true. If the target is _not_ a map the generated code will fail to compile.